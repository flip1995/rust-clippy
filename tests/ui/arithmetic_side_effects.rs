#![allow(
    clippy::assign_op_pattern,
    clippy::erasing_op,
    clippy::identity_op,
    clippy::unnecessary_owned_empty_strings,
    arithmetic_overflow,
    unconditional_panic
)]
#![feature(inline_const, saturating_int_impl)]
#![warn(clippy::arithmetic_side_effects)]

use core::num::{Saturating, Wrapping};

pub fn hard_coded_allowed() {
    let _ = 1f32 + 1f32;
    let _ = 1f64 + 1f64;

    let _ = Saturating(0u32) + Saturating(0u32);
    let _ = String::new() + "";
    let _ = Wrapping(0u32) + Wrapping(0u32);

    let saturating: Saturating<u32> = Saturating(0u32);
    let string: String = String::new();
    let wrapping: Wrapping<u32> = Wrapping(0u32);

    let inferred_saturating = saturating + saturating;
    let inferred_string = string + "";
    let inferred_wrapping = wrapping + wrapping;

    let _ = inferred_saturating + inferred_saturating;
    let _ = inferred_string + "";
    let _ = inferred_wrapping + inferred_wrapping;
}

#[rustfmt::skip]
pub fn non_overflowing_const_ops() {
    const _: i32 = { let mut n = 1; n += 1; n };
    let _ = const { let mut n = 1; n += 1; n };

    const _: i32 = { let mut n = 1; n = n + 1; n };
    let _ = const { let mut n = 1; n = n + 1; n };

    const _: i32 = { let mut n = 1; n = 1 + n; n };
    let _ = const { let mut n = 1; n = 1 + n; n };

    const _: i32 = 1 + 1;
    let _ = const { 1 + 1 };
}

pub fn non_overflowing_runtime_ops() {
    let mut _n = i32::MAX;

    // Assign
    _n += 0;
    _n -= 0;
    _n /= 99;
    _n %= 99;
    _n *= 0;
    _n *= 1;

    // Binary
    _n = _n + 0;
    _n = 0 + _n;
    _n = _n - 0;
    _n = 0 - _n;
    _n = _n / 99;
    _n = _n % 99;
    _n = _n * 0;
    _n = 0 * _n;
    _n = _n * 1;
    _n = 1 * _n;
    _n = 23 + 85;
}

#[rustfmt::skip]
pub fn overflowing_runtime_ops() {
    let mut _n = i32::MAX;

    // Assign
    _n += 1;
    _n -= 1;
    _n /= 0;
    _n %= 0;
    _n *= 2;

    // Binary
    _n = _n + 1;
    _n = 1 + _n;
    _n = _n - 1;
    _n = 1 - _n;
    _n = _n / 0;
    _n = _n % 0;
    _n = _n * 2;
    _n = 2 * _n;
}

fn main() {}
