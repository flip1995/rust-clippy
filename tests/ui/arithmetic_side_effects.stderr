error: arithmetic operation that can potentially result in unexpected side-effects
  --> $DIR/arithmetic_side_effects.rs:80:5
   |
LL |     _n += 1;
   |     ^^^^^^^
   |
   = note: `-D clippy::arithmetic-side-effects` implied by `-D warnings`

error: arithmetic operation that can potentially result in unexpected side-effects
  --> $DIR/arithmetic_side_effects.rs:81:5
   |
LL |     _n -= 1;
   |     ^^^^^^^

error: arithmetic operation that can potentially result in unexpected side-effects
  --> $DIR/arithmetic_side_effects.rs:82:5
   |
LL |     _n /= 0;
   |     ^^^^^^^

error: arithmetic operation that can potentially result in unexpected side-effects
  --> $DIR/arithmetic_side_effects.rs:83:5
   |
LL |     _n %= 0;
   |     ^^^^^^^

error: arithmetic operation that can potentially result in unexpected side-effects
  --> $DIR/arithmetic_side_effects.rs:84:5
   |
LL |     _n *= 2;
   |     ^^^^^^^

error: arithmetic operation that can potentially result in unexpected side-effects
  --> $DIR/arithmetic_side_effects.rs:87:10
   |
LL |     _n = _n + 1;
   |          ^^^^^^

error: arithmetic operation that can potentially result in unexpected side-effects
  --> $DIR/arithmetic_side_effects.rs:88:10
   |
LL |     _n = 1 + _n;
   |          ^^^^^^

error: arithmetic operation that can potentially result in unexpected side-effects
  --> $DIR/arithmetic_side_effects.rs:89:10
   |
LL |     _n = _n - 1;
   |          ^^^^^^

error: arithmetic operation that can potentially result in unexpected side-effects
  --> $DIR/arithmetic_side_effects.rs:90:10
   |
LL |     _n = 1 - _n;
   |          ^^^^^^

error: arithmetic operation that can potentially result in unexpected side-effects
  --> $DIR/arithmetic_side_effects.rs:91:10
   |
LL |     _n = _n / 0;
   |          ^^^^^^

error: arithmetic operation that can potentially result in unexpected side-effects
  --> $DIR/arithmetic_side_effects.rs:92:10
   |
LL |     _n = _n % 0;
   |          ^^^^^^

error: arithmetic operation that can potentially result in unexpected side-effects
  --> $DIR/arithmetic_side_effects.rs:93:10
   |
LL |     _n = _n * 2;
   |          ^^^^^^

error: arithmetic operation that can potentially result in unexpected side-effects
  --> $DIR/arithmetic_side_effects.rs:94:10
   |
LL |     _n = 2 * _n;
   |          ^^^^^^

error: aborting due to 13 previous errors

